### DISCUSSION TITLE
TagLab crashing when trying to use 4-clicks or Split screen

### DISCUSSION THREAD

**Discussion by JohnStratford** [cnr-isti-vclab/TagLab#120] [General]
Hi, 

Whenever I try to use the '4-clicks segmentation' tool or activate the 'Split screen' function' TagLab freezes for a few minutes then crashes and closes. When this happens, this error message is returned in the Windows Powershell I used to launch the software: 

**1. After trying to use '4-clicks segmentation':** 

PS C:\Users\Downloads\TagLab + required packages\TagLab-main> python.exe taglab.py
TagLab 2023.5.17
Traceback (most recent call last):
  File "C:\Users\Downloads\TagLab + required packages\TagLab-main\source\QtImageViewerPlus.py", line 627, in drawForeground
    painter.drawLine(self.mouseCoords.x(), rect.top(), self.mouseCoords.x(), rect.bottom())
TypeError: arguments did not match any overloaded call:
  drawLine(self, l: QLineF): argument 1 has unexpected type 'float'
  drawLine(self, line: QLine): argument 1 has unexpected type 'float'
  drawLine(self, x1: int, y1: int, x2: int, y2: int): argument 1 has unexpected type 'float'
  drawLine(self, p1: QPoint, p2: QPoint): argument 1 has unexpected type 'float'
  drawLine(self, p1: Union[QPointF, QPoint], p2: Union[QPointF, QPoint]): argument 1 has unexpected type 'float'

**2. After activating the 'Split screen' function':**

PS C:\Users\Downloads\TagLab + required packages\TagLab-main> python.exe taglab.py
TagLab 2023.5.17
Traceback (most recent call last):
  File "C:\Users\Downloads\TagLab + required packages\TagLab-main\source\QtImageViewer.py", line 132, in setViewParameters
    self.horizontalScrollBar().setValue(posx)
TypeError: setValue(self, a0: int): argument 1 has unexpected type 'float'

(I am using Windows 11, and I have installed 64-bit Python 3.10.11, CUDA Toolkit 11.6 and the x64 Visual Studio Redistributable.)
Does anyone know what is causing these errors and how I can fix them?
Thanks for your help!

**Comment by damarislanders** [comment]
I am getting this exact same crash and errors after trying to use '4-clicks segmentation'. This error is happening once a project with associated map (.jpg) and labels are loaded. The error is triggered as soon as the user presses the 4-clicks icon.

I tried editing the drawline() method, in the QtImageViewerPlus.py file, to explicitly accept floats or int in different combinations depending on the errors thrown:
`#used for crossair cursor
    def drawForeground(self, painter, rect):
        if self.showCrossair:
            painter.setClipRect(rect)
            pen = QPen(Qt.white, 1)
            pen.setCosmetic(True)
            painter.setPen(pen)
            #painter.drawLine(self.mouseCoords.x(), rect.top(), self.mouseCoords.x(), rect.bottom())
            #painter.drawLine(rect.left(), self.mouseCoords.y(), rect.right(), self.mouseCoords.y())
            painter.drawLine(int(self.mouseCoords.x()), float(rect.top()), int(self.mouseCoords.x()), float(rect.bottom()))
            painter.drawLine(float(rect.left()), int(self.mouseCoords.y()), float(rect.right()), int(self.mouseCoords.y()))
`

However, this is still throwing a similar error:

Traceback (most recent call last):
  File "c:\Users\user\Downloads\TagLab-main\source\QtImageViewerPlus.py", line 629, in drawForeground
    painter.drawLine(int(self.mouseCoords.x()), float(rect.top()), int(self.mouseCoords.x()), float(rect.bottom()))
TypeError: arguments did not match any overloaded call:
  drawLine(self, l: QLineF): argument 1 has unexpected type 'int'
  drawLine(self, line: QLine): argument 1 has unexpected type 'int'
  drawLine(self, x1: int, y1: int, x2: int, y2: int): argument 2 has unexpected type 'float'
  drawLine(self, p1: QPoint, p2: QPoint): argument 1 has unexpected type 'int'
  drawLine(self, p1: Union[QPointF, QPoint], p2: Union[QPointF, QPoint]): argument 1 has unexpected type 'int'

It might be that the values being passed into `self.mouseCoords.x(), rect.top(), self.mouseCoords.y(), and rect.bottom()` aren't explicitly set in the code. Do they come from another source? It may be worth investigating how they're calculated and, further to the error messages, making sure that they are the correct types before passing them to drawLine().

**Comment by maxcorsini** [comment]
Hi, 

the combination of Python 3.10.x and new version of Qt cause this issue. We can fix it but it requires a lot of work, so for the moment, I suggest to uninstall Python 3.10, install Python 3.9, and then re-install TagLab by running again install.py 

Please, let me know if this solve.

Best
Massimiliano

**Reply by damarislanders** [reply]
Hi @maxcorsini,

This worked! - I am now using Python 3.9 in my conda environment :)

Thank you very much.
My regards,
Damaris
